src/main/scala/com/fulcrumgenomics/util/Sorter.scala

```diff
@@ -0,0 +1,126 @@
+               sampleSize: Int         = DefaultSampleSize,
+               inflationFactor: Double = DefaultInflationFactor)(implicit tag: TypeTag[A]): Sorter[A] = {
+    new Sorter(codec, comparator, Some(memory), None, tmpDirs, sampleSize, inflationFactor)
+  }
+
+  /** Constructs a Sorter that attempts limits based on the number of records. */
+  def apply[A](codec: Codec[A],
+               comparator: Comparator[A],
+               records: Int,
+               tmpDirs: Seq[DirPath])(implicit tag: TypeTag[A]): Sorter[A] = {
+    val tmp = if (tmpDirs.isEmpty) DefaultTmpDirs else tmpDirs
+    new Sorter(codec, comparator, None, Some(records), tmp, 0, 1)
+  }
+}
+
+
+class Sorter[A] private (private val codec:   Codec[A],
+                         private val comparator: Comparator[A],
+                         private val memory:  Option[Long] = None,
+                         private val records: Option[Int] = None,
+                         val tmpDirs: Seq[DirPath] = Seq(Io.defaultTempDir()),
+                         val sampleSize: Int = 1000,
+                         val inflationFactor: Double = 5)(implicit tag: TypeTag[A]) extends Iterable[A] {
+
+  /** Class to help size up how big the records are. */
+  private class Sizer {
+    val buffer = new ListBuffer[A]()
+    val stream = new OutputStream {
+      var bytes = 0L
+      override def write(b: Int): Unit = bytes += 1
+    }
+    /** Calculates the average serialized size in bytes of the records sampled. */
+    def bytesPerRecord: Int = Math.round(Math.ceil(stream.bytes / buffer.size.toDouble)).toInt
+  }
+
+  /** Use reflection to grab the Class[A] for the record type. */
+  private val recordType = ReflectionUtil.typeTagToClass[A]
+
+  /** Use the supplied memory limit, or else half of the heap! */
+  private val memoryLimit = memory.getOrElse(Runtime.getRuntime.maxMemory() / 2)
+
+  /** Only one of these should be defined at a time. */
+  private var collection: Option[SortingCollection[A]] = None
+  private var sizer: Option[Sizer] = None
+
+  (memory, records) match {
+    case (Some(bs),     None) => this.sizer = Some(new Sizer)
+    case (None,     Some(rs)) => this.collection = Some(SortingCollection.newInstance(recordType, codec, comparator, rs))
+    case (_, _)               => throw new IllegalArgumentException("Must supply either memory or records but not both.")
+  }
+
+  override def iterator: Iterator[A] = (this.sizer, this.collection) match {
+    case (Some(s), None) => s.buffer.iterator
+    case (None, Some(c)) => c.iterator()
+    case _ => unreachable("Only sizer or collection should be defined at any one time.")
+  }
+
+  def +=(a: A): Unit = (this.sizer, this.collection) match {
+    case (None, Some(c)) =>
+      c.add(a)
+    case (Some(s), None) =>
+      s.buffer += a
+      this.codec.setOutputStream(s.stream)
+      this.codec.encode(a)
+
+      if (s.buffer.size >= this.sampleSize) {
+        System.gc()
```

Comment:
This is also here just for testing ... so that I can debug with a breakpoint here and see what the heap looks like immediately after GC'ing.